plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3AsyncDeleteHelper.java

```diff
@@ -0,0 +1,123 @@
+        S3BlobStore blobStore,
+        List<String> objectsToDelete,
+        CompletableFuture<Void> currentChain,
+        boolean ignoreIfNotExists,
+        Runnable afterDeleteAction
+    ) {
+        List<List<String>> batches = createDeleteBatches(objectsToDelete, blobStore.getBulkDeletesSize());
+        CompletableFuture<Void> newChain = currentChain.thenCompose(
+            v -> executeDeleteBatches(s3AsyncClient, blobStore, batches, ignoreIfNotExists)
+        );
+        if (afterDeleteAction != null) {
+            newChain = newChain.thenRun(afterDeleteAction);
+        }
+        return newChain;
+    }
+
+    static List<List<String>> createDeleteBatches(List<String> keys, int bulkDeleteSize) {
+        List<List<String>> batches = new ArrayList<>();
+        for (int i = 0; i < keys.size(); i += bulkDeleteSize) {
+            batches.add(keys.subList(i, Math.min(keys.size(), i + bulkDeleteSize)));
+        }
+        return batches;
+    }
+
+    private static CompletableFuture<Void> executeDeleteBatches(
+        S3AsyncClient s3AsyncClient,
+        S3BlobStore blobStore,
+        List<List<String>> batches,
+        boolean ignoreIfNotExists
+    ) {
+        CompletableFuture<Void> allDeletesFuture = CompletableFuture.completedFuture(null);
+
+        for (List<String> batch : batches) {
+            allDeletesFuture = allDeletesFuture.thenCompose(
+                v -> executeSingleDeleteBatch(s3AsyncClient, blobStore, batch, ignoreIfNotExists)
+            );
+        }
+
+        return allDeletesFuture;
+    }
+
+    private static CompletableFuture<Void> executeSingleDeleteBatch(
+        S3AsyncClient s3AsyncClient,
+        S3BlobStore blobStore,
+        List<String> batch,
+        boolean ignoreIfNotExists
+    ) {
+        DeleteObjectsRequest deleteRequest = bulkDelete(blobStore.bucket(), batch, blobStore);
+        return s3AsyncClient.deleteObjects(deleteRequest)
+            .thenApply(response -> processDeleteResponse(response, ignoreIfNotExists))
+            .exceptionally(e -> {
+                if (!ignoreIfNotExists) {
+                    throw new CompletionException(e);
+                }
+                logger.warn("Error during batch deletion", e);
+                return null;
+            });
+    }
+
+    private static Void processDeleteResponse(DeleteObjectsResponse deleteObjectsResponse, boolean ignoreIfNotExists) {
+        if (!deleteObjectsResponse.errors().isEmpty()) {
+            if (ignoreIfNotExists) {
+                logger.warn(
+                    () -> new ParameterizedMessage(
+                        "Failed to delete some blobs {}",
+                        deleteObjectsResponse.errors()
+                            .stream()
+                            .map(s3Error -> "[" + s3Error.key() + "][" + s3Error.code() + "][" + s3Error.message() + "]")
+                            .collect(Collectors.toList())
+                    )
+                );
```

Comment:
how is this logging related to `ignoreIfNotExists` ?

Reply:
Removed the ignoreIfNotExists as we have similar behaviour for `delete()` and `deleteBlobsIgnoringIfNotExists(..)`.